Feature/1.7/branch on inbound

lazan opened the pull request:
This is a huge PR. I'm not sure how we're going to review this...

lazan committed:
log: added log error/warning with error-code enums

This patch moved log::line to it's own header and added functions to log
error and warning with error-code enums, which is a common use case.

lazan committed:
service: renamed service call semantic 'forward' to 'no_reply'

no_reply describes exactly the intention. forward does not.

lazan committed:
transaction: made transaction::global::ID hashable

lazan committed:
transaction: adding message type - branch request

Adding message for branch request to TM.

lazan committed:
transaction: code - collapse namespace

lazan committed:
transaction: add functionality to branch inbound transaction request

issue #238

lazan committed:
transaction: strong typed for global and branch ranges

This patch adds strongly typed spans for global and branch, this makes
it harder to misuse global and branch ranges.

Also: Made transaction::ID dependent on global::ID, instead of the
other way around. Makes more sense that the whole trid knows about the
global part.

lazan committed:
sql: generalized parameter_bind overload for binary like range

lazan committed:
gateway: some tcp generalization

lazan committed:
gateway: branch transaction in inbound

This patch branches transactions on inbound. Asks TM for the branch for
at given gtrid. This gives several benefits: TM is responsible for the
branching. Several different transaction branches from other domains
will converge to one domain local transaction branch.

outbound still branches which will be removed in a later commit.

issue #238

lazan committed:
gateway: remove all transaction branching from outbound

This commit does not work. outbound fan out to all involved connections
for a given gtrid. TM might do a one-phase-commit-optimization which
outbound just pass along. For it to work outbound must do complicated
transaction coordination. For a OPCO it must first do a prepare, then
commit/rollback, hence do stuff that TM should do.

This will work out in the upcoming commit where every connection has
it's own ipc, and every connection act as a resource. Hence let TM do
all the orchestration.

lazan committed:
common: ability to pass extra arguments via message::dispatch

This patch adds the ability to define extra arguments to be passed to
handlers via message::dispatch::basic_handler.
With this, it gets a lot easier to pass what file-descriptor the message
was consumed from.
The code in the handler, and the surrounding state can be made simpler.

lazan committed:
gateway: pass tcp descriptor to handlers

This patch passes the tcp file-descriptor to handlers that are in need
of knowing which connection the message was consumed from.

lazan committed:
common: add strong ipc file-descriptor type

This patch adds a new strong _ipc-file-descriptor_ type. We need
different strong types for tcp and ipc descriptors to help us
differentiate between ipc and tcp "connections", to make it harder to
produce bugs.

lazan committed:
algorithm: improved a few things with concepts

lazan committed:
common: simplified comparison for strong::Type, Uiid and Process

Took advantage of c++20 auto symmetric generation for equality.

lazan committed:
service: use ipc id for remote instances

This is a preparation when gateway-outbound starts using a dedicated
ipc-device for each tcp-connection.

issue #238

lazan committed:
queue: prepare queue for outbound ipc-tcp pair

lazan committed:
gateway: strong type for socket descriptor and ipc descriptor

This patch uses strong types for "tcp" and "ipc" descriptors. This will
help us keep these two file descriptor types separated, hence harder to
produce bugs.

lazan committed:
common: generalized compare equal to process::Handle

A concept that answer if a type can compare equal to process::Handle.
With abbreviated function template it gets pretty nice.

lazan committed:
common: added template 'key' to task::concurrent for better correlation

This enables the task-units to be invoked with the 'key' as second
argument. The task-unit does not need to keep this in it's shared state.
Also gives the benefit of _failing_ a 'key'. Useful when the key is a
socket-descriptor and we've lost a connection.

lazan committed:
common: optional outbound device not to throw if not available

If an optional outbound device is not on-line/available it is just going
to fail a send with empty return correlation.

lazan committed:
common: add multiplex device api for sending an event

Before, we only had a blocking send.

lazan committed:
unittest: associate a unittest service-send with a transaction

lazan committed:
common: two new events. ipc_destroyed, transaction_disassociate

ipc_destroyed, when an ipc is destroyed.
transaction_disassociate, when TM thinks a trid is done

lazan committed:
transaction: add messages for asking if transactions are active or not

lazan committed:
common: add 'failed' for send::Coordinator

Enables explicit failed for a given ipc-address. This will find the
corresponding 'remote' and invoke the error callback, then delete the
'remote'.

lazan committed:
common: added log::verbose::error that log code to verbose::error

lazan committed:
common: conversation on service reply set duplex to 'terminated'

When a service that is invoked in conversation mode does a service
return, we set 'duplex' in the conversation::Send message to
'terminated' to indicate to caller that this conversation is
disconnected/terminated

lazan committed:
common: add reset indication to message::service::concurrent::Advertise

This to enable to tell SM that all services associated to the ipc is to
be removed/unadvertised.

issue #238

lazan committed:
common: some logging improvement

lazan committed:
discovery: provider registration to an ipc device

This to enable a process to register several inbound devices as
providers.

Also:
* make sure we handle ipc::Destroyed events.
* improved unittest abstraction

lazan committed:
domain: handle event ipc::Destroyed and transaction::Disassociate

Just as the old events, if there are subscribers -> sent the event to
them.

lazan committed:
transaction: use ipc-address as key for resources

Before: We used pid as the key for a given resource.
Now: We use ipc-address which make it possible for processes to act as
multiple resources to TM. Needed for gateway

issue #238

lazan committed:
service: pid -> ipc  & route calls to same instance based on gtrid

We now keep track of instances via their ipc-address instead of pid.
We also keep track of gtrid of each lookup request and to which
instance/ipc we reply with. This enables SM to let as much calls as
possible go to the same remote instance based on gtrid. That is, keep
the transaction as little distributed as possible.

issue #238

lazan committed:
queue: manager act on ipc-destroyed-event. use ipc as key

Before: We keept track of pid:s as keys to "addresses" for queues.

Now: We keep track of the `process` including ipc to find
"destroyed"/removed queues

Also added a `reset' to advertise message to indicate that all queues to
the ipc should be removed.

lazan committed:
queue: forward act on ipc-destroyed-event

We now act on the ipc-destroyed-event trying to invalidate/restart
relevant forwards.

Some improvements to the invalidation/restart. Should be more resilient.

lazan committed:
queue: group act on ipc-destroyed-event

We remove pending if the reply-address matches the destroyed ipc

lazan committed:
gateway: one inbound-ipc device for each tcp connection

This is a huge commit, and I couldn't easily split it to more bit size
chunks, unfortunately.

This patch using a dedicated ipc-device to each tcp connection. This
makes it a lot easier to deduce which "connection pair" messages should
be sent to.

outbound: Does not need to fan-out, and keep track of where services and
queues reside any more. It just advertise the services/queues for a
given ipc, and when messages arrives at this icp, we know what
tcp-connection we should send it to.

issue #238

lazan committed:
unittest: simplified a few unittests

lazan committed:
code: fix some g++11 concept related problems

lazan commented on middleware/common/include/casual/task/concurrent.h:
Hmm, this should not be in this commit...

lazan committed:
common: add a "concurrent" task abstraction

An abstraction that helps handle tasks that is meant to be executed in a
concurrent/parallel way. That is, any task could make progress, and be
done.

lazan committed:
common: pass through for byte size types in byteorder

AlbsAgren commented on middleware/common/include/common/log/line.h:
```diff
@@ -0,0 +1,46 @@
+//!
```
Neat!

AlbsAgren commented on middleware/common/include/common/message/transaction.h:
```diff
@@ -19,388 +20,399 @@
+      namespace coordinate::inbound
+      {
+         using base_request = message::basic_request< message::Type::transaction_coordinate_inbound_request>;
+         struct Request : base_request
```
This is to request a branch right? Isnt something like inbound::branch::Request like in the commit message more descriptive?

lazan replied:
I concur.

AlbsAgren commented on middleware/common/include/common/strong/span.h:
```diff
@@ -0,0 +1,38 @@
+         using base_type = std::span< T>;
+         using base_type::base_type;
+
+         //template< std::contiguous_iterator Iter>
```
Can this be removed?

AlbsAgren commented on thirdparty/database/include/sql/database.h:
```diff
@@ -149,6 +149,12 @@ namespace sql
          return code::make( sqlite3_bind_blob( statement, index, value.data(), value.size(), SQLITE_STATIC));
       }
 
+      template< casual::concepts::binary::like Range>
```
Dont know if it matters, but I'm guessing this change is supposed to be in [sql: generalized parameter_bind overload for binary like range](https://github.com/casualcore/casual/pull/329/commits/350e7320801ed017a9d3a54c7cf5f1e7ab06455c)?

lazan replied:
Yes, probably :)

lazan replied:
I'll let this slide...

AlbsAgren commented on middleware/common/include/common/strong/id.h:
```diff
@@ -55,9 +50,32 @@ namespace casual
       } // file
 
-      namespace socket
+      namespace ipc
```
This feels very helpful!

AlbsAgren commented on middleware/domain/source/discovery/state.cpp:
```diff
@@ -32,7 +32,7 @@ namespace casual
          void Providers::registration( const message::discovery::api::provider::registration::Request& message)
          {
             // we only add 'new' processes
```
Very nitpicky, but now we're adding devices right?

AlbsAgren commented on middleware/transaction/source/manager/handle.cpp:
```diff
@@ -400,35 +489,46 @@ namespace casual
+                        // some sanity checks
+                        {
+                           if( transaction->stage != state::transaction::Stage::prepare)
+                              common::log::error( common::code::casual::invalid_semantics, "transaction is not in prepare phase - actual: ", transaction->stage, " - action: ignore");
```
Should there be a return here seeing as the the log message says `action - ignore`?

lazan replied:
We might terminate. Something is wrong with our own code...

AlbsAgren commented on middleware/service/unittest/source/manager/test_state.cpp:
```diff
@@ -249,14 +249,14 @@ namespace casual
 
             // we remove the instance
-            service->remove( )
+            //service->remove( )
```
🧹

lazan replied:
Check.

AlbsAgren commented on middleware/queue/source/manager/state.cpp:
```diff
@@ -90,15 +90,37 @@ namespace casual
+      {
+         namespace
+         {
+            void remove_queues( State& state, common::process::compare_equal_to_handle auto id)
```
You can now 'constrain' auto? Fancy

lazan replied:
Abbreviated function template. I think it's really nice. A lot less verbose, and it gets pretty self explanatory with a concept.

AlbsAgren commented on middleware/queue/source/forward/handle.cpp:
```diff
@@ -926,25 +931,93 @@ namespace casual
+                           // if we've got pending enqueues to lost target we need to fake a reply
+                           // to keep the "state-machine" going
+                           for( auto& pending : algorithm::filter( state.pending.enqueues, is_forward( forward.id)))
+                              detail::push::faked::reply( pending);
```
This feels a bit unfortunate, but I can't think of a better way to do it

lazan replied:
I concur...

AlbsAgren commented on middleware/gateway/include/gateway/group/inbound/tcp.h:
```diff
@@ -21,6 +21,7 @@ namespace casual
          return group::tcp::send( state, &handle::connection::lost, descriptor, std::forward< M>( message));
       }
 
+/*
```
Can this be removed?

lazan replied:
Remove!

AlbsAgren commented on middleware/gateway/unittest/source/group/outbound/test_state.cpp:
```diff
@@ -15,11 +15,12 @@ namespace casual
 
    namespace gateway::group::outbound
    {
+      /*
```
Can this be removed?

lazan replied:
Check.

AlbsAgren reviewed (commented):
Feels like a major improvement overall!

lazan commented on middleware/gateway/include/gateway/group/connection.h:
Remove, or change external::tcp... to this...

lazan commented on middleware/gateway/include/gateway/group/inbound/state.h:
```diff
@@ -339,8 +212,8 @@ namespace casual
-            CASUAL_SERIALIZE( correlations);
-            CASUAL_SERIALIZE( conversations);
+            //CASUAL_SERIALIZE( correlations);
+            //CASUAL_SERIALIZE( conversations);
```
Remove

OlofGullnas replied:
There is another case of a commented CASUAL_SERIALIZE a bit up in the file (around line 174)

lazan commented on middleware/gateway/source/group/connection.cpp:
Remove, or move tcp to this.

lazan commented on middleware/gateway/source/group/inbound/task/create.cpp:
```diff
@@ -215,82 +213,161 @@ namespace casual
 
                   },
-                  [ &state, shared]( common::message::transaction::coordinate::inbound::Reply& reply) mutable
+                  [ &state, shared]( common::message::transaction::coordinate::inbound::Reply& reply, strong::socket::id descriptor) mutable
```
Change name to transaction::branch::Request... ?

lazan commented on middleware/gateway/source/group/outbound/handle.cpp:
```diff
@@ -40,133 +40,77 @@ namespace casual
+               namespace precondition
+               {
+                  bool reply( State& state, strong::socket::id descriptor, auto& message, auto&& create_reply)
+                  {
```
Better name?

lazan replied:
Future us will decide...

lazan commented on middleware/gateway/source/group/outbound/handle.cpp:
```diff
@@ -717,159 +555,118 @@ namespace casual
-                     } // prepare
+                           auto destination = state.reply_destination.extract( message.correlation);
+                           state.multiplex.send( destination.ipc, message);
+                        };
```
Why not use the transaction dissasociate::transaction event?

OlofGullnas commented on middleware/queue/source/forward/handle.cpp:
```diff
@@ -926,25 +931,93 @@ namespace casual
+                        log::line( verbose::log, "state.pending: ", state.pending);
+
+                        static constexpr auto is_forward = []( auto id) { return [ id]( auto& forward){ return forward == id;};};
+
```
I built this branch with g++-13 (with my changes for gcc-13 feature/1.7/main rebased to this branch). 
I got a warning/error:
`compile: /export/home/guo/c_1.7/casual-1.7/middleware/queue/source/forward/handle.cpp
error:
/export/home/guo/c_1.7/casual-1.7/middleware/queue/source/forward/handle.cpp: In instantiation of ‘void casual::queue::forward::local::{anonymous}::handle::dead::detail::dead_event(casual::queue::forward::State&, auto:153) [with auto:153 = casual::common::strong::Type<int, casual::common::strong::detail::integral::policy<int, casual::common::strong::process::tag, -1> >]’:
/export/home/guo/c_1.7/casual-1.7/middleware/queue/source/forward/handle.cpp:1029:43:   required from here
/export/home/guo/c_1.7/casual-1.7/middleware/queue/source/forward/handle.cpp:964:47: error: variable ‘is_forward’ set but not used [-Werror=unused-but-set-variable]
  964 |                         static constexpr auto is_forward = []( auto id) { return [ id]( auto& forward){ return forward == id;};};
      |                                               ^~~~~~~~~~
/export/home/guo/c_1.7/casual-1.7/middleware/queue/source/forward/handle.cpp: In instantiation of ‘void casual::queue::forward::local::{anonymous}::handle::dead::detail::dead_event(casual::queue::forward::State&, auto:153) [with auto:153 = casual::common::strong::Type<casual::common::Uuid, casual::common::strong::ipc::tag>]’:
/export/home/guo/c_1.7/casual-1.7/middleware/queue/source/forward/handle.cpp:1041:43:   required from here
/export/home/guo/c_1.7/casual-1.7/middleware/queue/source/forward/handle.cpp:964:47: error: variable ‘is_forward’ set but not used [-Werror=unused-but-set-variable]
cc1plus: all warnings being treated as errors

aborting due to errors`
I can't say if this is a legitimate warning or not. I looked arounf a little in bugzilla for gcc/g++ and there seemed to be som issues having to do with  constexpr and parameters and lambda captures in the gcc-9 to gcc-11 timeframe that has been corrected.
I removed static and added "explicit" capture of "is_forward" to the invalidate_queue_forward and invalidate_service_forward lambdas. Then code builds without warings. And unit tests run.

lazan replied:
You should be able to create a constexpr-static non capturing lambda and it behaves just as a free function. Maybe it needs to be `inline` ?

OlofGullnas replied:
I (!) think it is a compiler bug. Turning off the warning is a possibility, but it may be useful and correct in other cases. As far as I know this is the only place the warning is produced with casual.

For some reason the compiler misses that the variable "static constexpr is_forward" is used by  the lambdas invalidate_queue_forward and invalidate_service_forward. Some of the problems they fixed in earlier g++ versions seemed to have to do with the compiler missing to "mark" the use of some variables on some paths through the compiler. There might be one more for the case when a static constexpr lambda is used by another lambda... I will try to reproduce in a small code example. Perhaps not easy. I do not know  how much context is needed...

lazan replied:
Try to use it without the constexpr. That should also work, but then the compiler needs to guard the initialization for concurrent threads.

OlofGullnas replied:
Already tried. Same compiler warning. Turning off warning or remove static and capture (by value is what I tried) is_forward variable seems to work.

lazan replied:
High probability for a compiler bug. I'll say we lay low with g++13 for a while. At least until we chose to use it in our build pipeline.

lazan committed:
transaction: message name coordinate::inbound:: -> inbound::branch::

Rename `transaction::coordinate::inbound::Request/Reply` to
`transaction::inbound::branch::Request/Reply`

lazan committed:
code: removed obsolete/commented code

OlofGullnas commented on middleware/common/include/common/network/byteorder.h:
```diff
@@ -29,18 +30,12 @@ namespace casual
-               struct transcode<char>
+               // just passthrough 
+               template< typename T> requires concepts::any_of< T, bool, char, signed char, unsigned char>
+               struct transcode< T> 
```
Should this (for completeness since c++20) include char8_t in the "any_of" list of types?

In C char8_t is a "typedef char8_t unsined char", but I think it is a distinct type in C++ 20.
 I don't think char8_t is used in casual so this comment is somewhat academical...

lazan replied:
I concur. And char8_t should be used :) But I think we do this when (if) we use std::u8string "everywhere". It's up for debate... @KristianIvarsson and I have talked a bit about it. We should take a stance which "encoding" we're handling/supporting (now we don't "care", and the user needs to have the same encoding on all machines). If we take a stance, then utf8 it is -> std::u8string "everywhere".

lazan committed:
gateway: renamed/moved 'tcp::External' to 'connection::Handle'

Since we now have tcp-ipc-pair as a 'connection', the name _external_ is
misleading.

This patch moved the "external holder" to `group/connection.h´ with name
`group::connection::Holder` which is (arguable) a much better name.

lazan committed:
common: improved on transaction::global::ID

* implicit conversion to range-type
* transaction::ID takes global::id::range instead of global::ID

lazan committed:
transaction: always send event when transactions are removed

Removed the _inbound-branch_ lookup state and just use the regular
transaction state.

When a transaction is removed, we always send a disassociate event.

lazan committed:
gateway: outbound listen to event::transaction::Disassociate

Use the event to remove state that outbound needs to know if there are
in-flight transaction for a given connection

AlbsAgren reviewed (approved):
Looks real good I think

lazan commented:
Merged manually
